shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java

```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
+            long requestId = in.readLong();
+            String segment = in.readUTF();
+            return new TableKeysRemoved(requestId, segment);
+        }
+    }
+
+    @Data
+    public static final class ReadTable implements Request, WireCommand {
+
+        final WireCommandType type = WireCommandType.READ_TABLE;
+        final long requestId;
+        final String segment;
+        final String delegationToken;
+        final List<TableKey> keys; // the version of the key is always set to io.pravega.segmentstore.contracts.tables.TableKey.NO_VERSION
+
+        @Override
+        public void process(RequestProcessor cp) {
+            cp.readTable(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeUTF(delegationToken);
+            out.writeInt(keys.size());
+            for (TableKey key : keys) {
+                key.writeFields(out);
+            }
+        }
+
+        public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOException {
+            long requestId = in.readLong();
+            String segment = in.readUTF();
+            String delegationToken = in.readUTF();
+            int numberOfKeys = in.readInt();
+            List<TableKey> keys = new ArrayList<>(numberOfKeys);
+            for (int i = 0; i < numberOfKeys; i++) {
+                keys.add((TableKey) TableKey.readFrom(in, in.available()));
+            }
+            return new ReadTable(requestId, segment, delegationToken, keys);
+        }
+    }
+
+    @Data
+    public static final class TableRead implements Reply, WireCommand {
+        final WireCommandType type = WireCommandType.TABLE_READ;
+        final long requestId;
+        final String segment;
+        final TableEntries entries;
+
+        @Override
+        public void process(ReplyProcessor cp) {
+            cp.tableRead(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            entries.writeFields(out);
+        }
+
+        public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOException {
+            long requestId = in.readLong();
+            String segment = in.readUTF();
+            TableEntries entries = (TableEntries) TableEntries.readFrom(in, in.available());
+            return new TableRead(requestId, segment, entries);
+        }
+    }
+
+    @Data
+    public static final class TableEntries implements WireCommand {
+        final WireCommandType type = WireCommandType.TABLE_ENTRIES;
+        final List<Map.Entry<TableKey, TableValue>> entries;
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeInt(entries.size());
+            for (Map.Entry<TableKey, TableValue> ent :entries) {
```

Comment:
The way these table keys and values are being written makes it impossible to add more fields to table key and value in the future because one will end and any unread bytes will be assumed to be part of the next item in entries.

Reply:
Added payload size field to handle future updates to TableKey and TableValue.